### PR TITLE
openssl: use OPENSSL_init_ssl() with >= 1.1.0

### LINKED ISSUE #6254: Twice initialization openssl
### I did this

Full steps to reproduce problem:

1. Build GOST engine for openssl 1.1.1

```
$ mkdir /tmp/testgostengine/
$ git clone -b openssl_1_1_1 --single-branch https://github.com/gost-engine/engine /tmp/testgostengine/
$ mkdir /tmp/testgostengine/build/; cd /tmp/testgostengine/build/
$ cmake --configure .
$ cmake --build . --config Release
```

2. Write minimal config

```
$ cat /tmp/testgostengine/example.conf | sed 's|dynamic_path = ./bin/gost.so|dynamic_path = /tmp/testgostengine/bin/gost.so|g' > /tmp/testgostengine/openssl.conf
```

3. Use it to connect to test host
```
$ OPENSSL_CONF=/tmp/testgostengine/openssl.conf curl --insecure --head https://tlsgost-2001.cryptopro.ru/
```
Output of command

```
$ OPENSSL_CONF=/tmp/testgostengine/openssl.conf curl --insecure 'https://tlsgost-2001.cryptopro.ru/'
GOST engine already loaded
<!DOCTYPE html>
<html>
<head>
<title>Welcome to Crypto-Pro tls server!</title>
<style>
    body {
        width: 35em;
        margin: 0 auto;
        font-family: Tahoma, Verdana, Arial, sans-serif;
    }
</style>
</head>
<body>
<h1>Welcome to Crypto-Pro tls server!</h1>
<p> TLS connection with id-GostR3410-2001-CryptoPro-XchA-ParamSet no auth requred.</p>


</body>
</html>
```

And in this output we have some problem like *GOST engine already loaded*. Is it problem with twice initialization openssl in curl? For example, if I don't use curl and do fetch by only php and openssl like this, I haven't this error:

Test php script:

```
$ cat << 'EOF' > /tmp/testgostengine/test.php
<?php

$mySSLContext = array(
    "ssl" => array(
        "verify_peer" => false,
        "verify_peer_name" => false,
    ),
);

print_r(file_get_contents("https://tlsgost-2001.cryptopro.ru/", false, stream_context_create($mySSLContext)));
EOF
```

Output:

```
$ OPENSSL_CONF=/tmp/testgostengine/openssl.conf php /tmp/testgostengine/test.php
<!DOCTYPE html>
<html>
<head>
<title>Welcome to Crypto-Pro tls server!</title>
<style>
    body {
        width: 35em;
        margin: 0 auto;
        font-family: Tahoma, Verdana, Arial, sans-serif;
    }
</style>
</head>
<body>
<h1>Welcome to Crypto-Pro tls server!</h1>
<p> TLS connection with id-GostR3410-2001-CryptoPro-XchA-ParamSet no auth requred.</p>


</body>
</html>
```

No error *GOST engine already loaded*.

### I expected the following

```
$ OPENSSL_CONF=/tmp/testgostengine/openssl.conf curl --insecure 'https://tlsgost-2001.cryptopro.ru/'
<!DOCTYPE html>
<html>
<head>
<title>Welcome to Crypto-Pro tls server!</title>
<style>
    body {
        width: 35em;
        margin: 0 auto;
        font-family: Tahoma, Verdana, Arial, sans-serif;
    }
</style>
</head>
<body>
<h1>Welcome to Crypto-Pro tls server!</h1>
<p> TLS connection with id-GostR3410-2001-CryptoPro-XchA-ParamSet no auth requred.</p>


</body>
</html>
```

### curl/libcurl version

```
$ curl -V
curl 7.73.0 (x86_64-pc-linux-gnu) libcurl/7.73.0 OpenSSL/1.1.1h zlib/1.2.11 zstd/1.4.5 libidn2/2.3.0 libpsl/0.21.1 (+libidn2/2.3.0) libssh2/1.9.0 nghttp2/1.41.0
Release-Date: 2020-10-14
Protocols: dict file ftp ftps gopher http https imap imaps mqtt pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS GSS-API HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets zstd
```

### operating system

Archlinux
```
uname -a
Linux myzion 5.9.10-arch1-1 #1 SMP PREEMPT Sun, 22 Nov 2020 14:16:59 +0000 x86_64 GNU/Linux
```

### LINKED ISSUE #6256: Crash when curl is used from inside an OpenSSL engine
Hi!
The yubihsm-pkcs11 module internally uses the curl library, which on initialization calls CONF_modules_load_file to disable errors on missing config files (I presume). When yubihsm-pkcs11 is used by the openssl command through libengine-pkcs11-openssl on Ubuntu Focal or Groovy there is a crash in int_ctx_new (crypto/evp/pmeth_lib.c:171). This seems to happen because CONF_modules_load_file seems to be called twice in this case, once implicitly from openssl library initialization, and then again later when the pkcs11 engine initializes yubihsm-pkcs11, which calls curl_global_init, which in turn calls CONF_modules_load_file. Both calls use NULL as the config file name so they end up loading the default config file twice, and some debugging revealed that the EVP_PKEY_METHOD it finds contains an invalid function pointer in pmeth->init. It might be the fact that the pkcs11 engine gets loaded twice, but the second time it gets rejected because it is already registered with the same id, and subsequently freed. Maybe this freeing messes up some common state also used by the existing pkcs11 engine.

The command i use is

YUBIHSM_PKCS11_CONF=yubihsm_pkcs11.conf OPENSSL_CONF=engine.cnf openssl dgst -engine pkcs11 -keyform engine -sha256 -sign -label_xxx

I explicitly point to the following config file 'engine.cnf'

...
openssl_conf = openssl_init

[openssl_init]
engines = engines_section

[engines_section]
pkcs11 = pkcs11_section

[pkcs11_section]
engine_id = pkcs11
dynamic_path = /usr/lib/x86_64-linux-gnu/engines-1.1/pkcs11.so
MODULE_PATH = /usr/local/lib/pkcs11/yubihsm_pkcs11.so
PIN = 0001password
init = 0
...

And yubihsm_pkcs11.conf:

...
connector=http://localhost:12345
...

The line number comes from an openssl built with --debug from github branch OpenSSL_1_1_1-stable and installed to the default location /usr/local/lib (where it will be used by default).

The same crash happens in the openssl version that comes with Ubuntu Focal and Groovy, but not earlier Ubuntu releases.

### PR DESCRIPTION
**Pull request by bagder**
Reported-by: Kovalkov Dmitrii and Per Nilsson
Fixes #6254
Fixes #6256
Closes #6260

### CODE REVIEW

File: lib/vtls/openssl.c [hunk 1]
```diff
@@ -90,6 +90,10 @@
 #include <openssl/engine.h>
 #endif
 
+#if (OPENSSL_VERSION_NUMBER >= 0x10100000L) /* 1.1.0 or later */
```
Comment (jay) [inline lib/vtls/openssl.c:93]:
```suggestion
#if (OPENSSL_VERSION_NUMBER >= 0x10101000L) /* 1.1.1 or later */
```
https://github.com/openssl/openssl/issues/13548#issuecomment-734766906

**Review by jay — CHANGES_REQUESTED** [review]

File: lib/vtls/openssl.c (hunk 1 above)
Comment (bagder) [reply lib/vtls/openssl.c:93]: Oops...

**Comment by jay** [comment]
It looks like ossl 1.1.0 is still an open question in that OpenSSL bug. I tested with 1.1.0l and 1.1.1h and it's only called in the latter, even though they share similar initialization code. At some point there's a difference but I don't know where.

**Comment by bagder** [comment]
Keeping the fix for >= 1.1.1 then seems like the more careful approach so let's start with that....

File: lib/vtls/openssl.c [hunk 2]
```diff
@@ -1119,6 +1124,10 @@ static int x509_name_oneline(X509_NAME *a, char *buf, size_t size)
  */
 static int Curl_ossl_init(void)
 {
+#if (OPENSSL_VERSION_NUMBER >= 0x10100000L) &&  \
+  !defined(LIBRESSL_VERSION_NUMBER)
+  OPENSSL_init_ssl(INITFLAGS, NULL);
```
Comment (jay) [inline lib/vtls/openssl.c:1129]: There may be things we are currently doing here that are not done by [OPENSSL_init_ssl](https://www.openssl.org/docs/man1.1.0/man3/OPENSSL_init_crypto.html). We use ENGINE_load_builtin_engines and so the equivalent of that might be OPENSSL_INIT_ENGINE_ALL_BUILTIN. They document which flags are ignored for subsequent calls (if we're not the first library to call) and that's not one of them. Ok fine. But what if another library first initializes with OPENSSL_INIT_NO_ADD_ALL_CIPHERS, which is ignored for subsequent calls, then we would not have those ciphers available if I understand this right.

Comment (bagder) [reply lib/vtls/openssl.c:1129]: Good points. But I don't know what we can do about those doubts and unanswered questions. It seems to me that `OPENSSL_init_ssl` is now the mandated and blessed way to init the library and I would imagine that all other ways also risk getting some init glitches.

Comment (qpernil) [reply lib/vtls/openssl.c:1129]: I looked in the code for openssl 1.1.1 STABLE and ENGINE_load_builtin_engines  is just a call to OPENSSL_init_crypto(OPENSSL_INIT_ENGINE_ALL_BUILTIN, NULL) so maybe that flag should be added to curl's call ?

The openssl command's main() calls OPENSSL_init_ssl(OPENSSL_INIT_ENGINE_ALL_BUILTIN | OPENSSL_INIT_LOAD_CONFIG, NULL).

On the other hand it is probably a bad idea to call ENGINE_load_builtin_engines on openssl >= 1.1.0 as that would probably 'use up' the one chance yo seem to have to call OPENSSL_init_ssl / OPENSSL_init_crypto.

Comment (bagder) [reply lib/vtls/openssl.c]: Good find. I've added `OPENSSL_INIT_ENGINE_ALL_BUILTIN` to the bitmask now.

File: lib/vtls/openssl.c [hunk 3]
```diff
@@ -1115,6 +1120,10 @@ static int x509_name_oneline(X509_NAME *a, char *buf, size_t size)
  */
 static int Curl_ossl_init(void)
 {
+#if (OPENSSL_VERSION_NUMBER >= 0x10100000L) &&  \
+  !defined(LIBRESSL_VERSION_NUMBER)
+  OPENSSL_init_ssl(OPENSSL_INIT_ENGINE_ALL_BUILTIN | INITFLAGS, NULL);
```
Comment (jay) [inline lib/vtls/openssl.c:1125]: Fine with me. Consider moving all of it here, for example I think this is faster to understand
```suggestion
  uint64_t flags = OPENSSL_INIT_ENGINE_ALL_BUILTIN |
#ifdef CURL_DISABLE_OPENSSL_AUTO_LOAD_CONFIG
                   OPENSSL_INIT_NO_LOAD_CONFIG;
#else
                   OPENSSL_INIT_LOAD_CONFIG;
#endif
  OPENSSL_init_ssl(flags, NULL);
```

Comment (bagder) [reply lib/vtls/openssl.c:1125]: Seems fine to me, and do it `const`.

File: lib/vtls/openssl.c [hunk 4]
```diff
@@ -1122,7 +1117,14 @@ static int Curl_ossl_init(void)
 {
 #if (OPENSSL_VERSION_NUMBER >= 0x10100000L) &&  \
   !defined(LIBRESSL_VERSION_NUMBER)
-  OPENSSL_init_ssl(OPENSSL_INIT_ENGINE_ALL_BUILTIN | INITFLAGS, NULL);
+  const int64_t flags = OPENSSL_INIT_ENGINE_ALL_BUILTIN |
+#ifdef CURL_DISABLE_OPENSSL_AUTO_LOAD_CONFIG
+    OPENSSL_INIT_NO_LOAD_CONFIG
+#else
+    OPENSSL_INIT_LOAD_CONFIG
+#endif
```
Comment (bagder) [inline lib/vtls/openssl.c:1125]: BoringSSL didn't approve:

~~~
vtls/openssl.c:1120:25: error: ‘OPENSSL_INIT_ENGINE_ALL_BUILTIN’ undeclared (first use in this function); did you mean ‘OPENSSL_INIT_ADD_ALL_CIPHERS’?
   const int64_t flags = OPENSSL_INIT_ENGINE_ALL_BUILTIN |
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                         OPENSSL_INIT_ADD_ALL_CIPHERS

**Review by jay — APPROVED** [review]
flags type is uint64_t

> error: ‘OPENSSL_INIT_ENGINE_ALL_BUILTIN’ undeclared

makes sense, OPENSSL_load_builtin_modules does nothing in boringssl

**Pull request closed without merging**